integer/src/lib.rs

```diff
@@ -665,6 +665,36 @@ impl_integer_for_usize!(u32, test_integer_u32);
 impl_integer_for_usize!(u64, test_integer_u64);
 impl_integer_for_usize!(usize, test_integer_usize);
 
+/// Calculate r * a / b, avoiding overflows and fractions.
+fn multiply_and_divide<T: Integer + Clone>(r: T, a: T, b: T) -> T {
+    // See http://blog.plover.com/math/choose-2.html for the idea.
+    let g = gcd(r.clone(), b.clone());
+    (r/g.clone()) * (a / (b/g))
+}
+
+/// Calculate the binomial coefficient.
+pub fn binomial<T: Integer + Clone>(mut n: T, k: T) -> T {
+    // See http://blog.plover.com/math/choose.html for the idea.
+    if k > n {
+        return T::zero();
+    }
+    let two = T::one() + T::one();
+    if k > n.clone()/two {
```

Comment:
It's probably cheaper in general to compare `k > n - k`, so you can avoid a division.  Primitive integers should optimize divide-by-two to a right shift, but for bigint division is much more expensive than subtraction.  Too bad that `Integer` doesn't include `Shr`.